### DISCUSSION TITLE
Big update of group part

### DISCUSSION THREAD

**Discussion by Deopster** [Deopster/Mercenaries-Hearthstone-game-bot#12] [General]
1. Group function has been completely rewritten:
- the algorithm works 2.5 times faster
- small bugs removed
- now, when checking the existing group, it passes through all 6 heroes, checking whether the 3 first ones are those indicated in the files and whether the remaining 30 LVL have reached
- made a reserve for the future with the ability to choose how many supporting heroes you want to add 1 2 or 3
- now if one of the heroes of the group is not deleted and then reassembled, then this hero is simply deleted and replaced by another.

2. Has been done work with the pictures search algorithm:

- AutoScale Images Added
- now you don't need 4 screens to find a hero, one is enough, which speeds up work a lot.
- work has begun on auto-scaling under different Resolutions 

### Before
![image](https://user-images.githubusercontent.com/68296704/138873215-406890ca-e1d2-4449-b804-634e75fb1ec1.png)
### After
![image](https://user-images.githubusercontent.com/68296704/138873103-b675eebb-0613-42ee-ab38-b514d9d71429.png)

<hr /><em>This discussion was created from the release <a href='https://github.com/Deopster/Mercenaries-Hearthstone-game-bot/releases/tag/v0.1.2-beta'>Big update of group part</a>.</em>

**Comment by xu001186** [comment]
I am working on the similar project , from my view , it's better to use the contour instead of the matchtemplate in some scenarios.
Below is a example to get the ability list from hero:
![ability_after](https://user-images.githubusercontent.com/39233649/139072380-d9a6526e-e246-4008-8eec-1b96278e4404.png)

This is the example to list the available ability list only:
![ability_another](https://user-images.githubusercontent.com/39233649/139073141-bdf674e9-048e-422e-ae53-f4f4ca494661.png)

Example to list all the heroes and bosses by the contour
![heros](https://user-images.githubusercontent.com/39233649/139072723-ac17f30a-ecc9-4321-af43-9861a10339cc.png)

**Reply by Deopster** [reply]
I disagree absolutely since this will greatly simplify the PVP system and make it less flexible, and here are three reasons.
1. Now Bot determines the colors of my heroes and opponents, if you simplify the algorithm to determine the contours, he will not be able to give the maximum damage by attacking red - green blue - red green-blue.
2. Since we can't determine what kind of hero is on the board, we can't make an attack algorithm. For example, now Milhouse is first hit with a split, then with a single shot, and then arrows. But it won't be possible if we just focus on borders.
3. I don't understand why we need this? The algorithm will work faster for a maximum of a second, reducing the number of pictures? So I already did this by reducing the number of necessary screenshots of the heroes to 1.
